File: krono_core_frontend_movil/lib/global/utils/session.dart
/// =============================================================================
/// session.dart — Estado global de autenticación (singleton) con JWT.
///
/// GESTIONA:
/// - Usuario actual, Access Token y Refresh Token en memoria
/// - Login con JWT (loginWithJwt)
/// - Refresco automático del token (refreshAccessToken)
/// - Soporte multi-empresa para admin (adminEmpresaId)
/// - Notificación de cambios mediante ChangeNotifier
///
/// Singleton: SessionController.instance accede desde toda la app.
/// GoRouter usa refreshListenable para reaccionar a cambios de sesión.
/// =============================================================================

import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import '../models/models.dart';

/// CONTROLADOR: SessionController
/// Gestiona el estado de la sesión del usuario de forma global (Singleton).
/// Soporta JWT (Access + Refresh Token) para autenticación segura.
class SessionController extends ChangeNotifier {
  static final SessionController instance = SessionController._();
  SessionController._();

  Usuario? _user;
  String? _accessToken;
  String? _refreshToken;
  int? _adminEmpresaId;

  Usuario? get user => _user;
  String? get accessToken => _accessToken;
  String? get refreshToken => _refreshToken;
  bool get isAuth => _user != null;
  int? get adminEmpresaId => _adminEmpresaId;

  void setAdminEmpresaId(int? id) {
    _adminEmpresaId = id;
    notifyListeners();
  }

  /// Inicia sesión con JWT desde la respuesta del backend.
  /// Guarda el usuario y los tokens en memoria.
  void loginWithJwt({
    required String accessToken,
    required String refreshToken,
    required String id,
    required String nombre,
    required String email,
    required UserRole rol,
    String? empresaNombre,
    int? empresaId,
    String? avatar,
  }) {
    _accessToken = accessToken;
    _refreshToken = refreshToken;
    _user = Usuario(
      id: id,
      nombre: nombre,
      email: email,
      rol: rol,
      empresaNombre: empresaNombre,
      empresaId: empresaId,
      avatar: avatar,
    );
    notifyListeners();
  }

  /// Refresca el Access Token usando el Refresh Token.
  /// Se llama automáticamente desde ApiClient al recibir 401.
  Future<void> refreshAccessToken() async {
    if (_refreshToken == null) throw Exception('No hay refresh token');

    const String baseUrl =
        'https://kronobackend-ctach7cya4cxb9d8.spaincentral-01.azurewebsites.net/api';

    try {
      final response = await http.post(
        Uri.parse('$baseUrl/auth/refresh'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'refreshToken': _refreshToken}),
      );
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        _accessToken = data['accessToken'] as String?;
        _refreshToken = data['refreshToken'] as String?;
      } else {
        logout();
        throw Exception('Sesión expirada');
      }
    } catch (e) {
      logout();
      rethrow;
    }
  }

  /// Login legacy (sin JWT, para modo demo y compatibilidad).
  void login({
    required String email,
    required UserRole rol,
    String? nombre,
    String? id,
    String? empresaNombre,
    int? empresaId,
    String? avatar,
  }) {
    _user = Usuario(
      id: id ?? 'u-${rol.name}',
      nombre:
          nombre ??
          switch (rol) {
            UserRole.admin => 'Admin Kronos',
            UserRole.jefe => 'Jefa de salon',
            UserRole.empleado => 'Empleado Demo',
            UserRole.cliente => 'Cliente Demo',
          },
      email: email,
      rol: rol,
      empresaNombre: empresaNombre,
      empresaId: empresaId,
      avatar: avatar,
    );
    notifyListeners();
  }

  void logout() {
    _user = null;
    _accessToken = null;
    _refreshToken = null;
    notifyListeners();
  }
}

File: kronoCore_Frontend_Web/global/js/session.js
/*
 * session.js — Control de sesión de usuario en el navegador.
 * 
 * GESTIONA:
 * - Inicio de sesión con JWT (loginWithJwt)
 * - Persistencia en sessionStorage (se mantiene al recargar)
 * - Renovación automática del Access Token (refreshAccessToken)
 * - Cierre de sesión (logout)
 * - Notificación de cambios (evento session-changed)
 * - Soporte multi-empresa para admin (adminEmpresaId)
 * 
 * USO: SessionController.instance.loginWithJwt(data)
 */

/**
 * session.dart → JavaScript
 * Sesión global con JWT para auth + filtrado por rol.
 * Singleton: solo existe una instancia en toda la aplicación.
 */
class SessionController {
  static instance = new SessionController();

  constructor() {
    this._user = null;
    this._accessToken = null;
    this._refreshToken = null;
    this._load(); // Carga la sesión guardada al iniciar
  }

  get user() { return this._user; }
  get accessToken() { return this._accessToken; }
  get refreshToken() { return this._refreshToken; }
  get isAuth() { return this._user !== null; }

  /** Para que el admin pueda seleccionar una empresa específica */
  get adminEmpresaId() {
    const raw = sessionStorage.getItem('krono_admin_empresa_id');
    return raw ? Number(raw) : null;
  }

  setAdminEmpresaId(id) {
    if (id === null || id === undefined || id === '') {
      sessionStorage.removeItem('krono_admin_empresa_id');
    } else {
      sessionStorage.setItem('krono_admin_empresa_id', String(id));
    }
    window.dispatchEvent(new Event('session-changed'));
  }

  /**
   * Inicia sesión con la respuesta del backend (JWT + datos de usuario).
   * Guarda tokens y datos del usuario en memoria y sessionStorage.
   * 
   * @param {Object} data - Respuesta del endpoint /api/auth/login
   * @param {string} data.accessToken - Token JWT de acceso (1h)
   * @param {string} data.refreshToken - Token JWT de refresco (7d)
   * @param {string} data.id - ID del usuario
   * @param {string} data.nombre - Nombre del usuario
   * @param {string} data.email - Email del usuario
   * @param {string} data.rol - Rol del usuario (admin, jefe, empleado, cliente)
   * @param {string} data.empresaId - ID de la empresa (null para admin)
   */
  loginWithJwt(data) {
    this._accessToken = data.accessToken;
    this._refreshToken = data.refreshToken;
    this._user = {
      id: data.id,
      nombre: data.nombre,
      email: data.email,
      rol: data.rol,
      empresaNombre: data.empresaNombre || null,
      empresaId: data.empresaId != null ? Number(data.empresaId) : null,
    };
    this._save();
    window.dispatchEvent(new Event('session-changed'));
  }

  /**
   * Refresca el Access Token usando el Refresh Token.
   * Se llama automáticamente cuando una petición recibe 401.
   * 
   * FLUJO:
   * 1. Envía POST /api/auth/refresh con el Refresh Token
   * 2. Recibe nuevo Access Token y Refresh Token
   * 3. Actualiza tokens en memoria y sessionStorage
   */
  async refreshAccessToken() {
    if (!this._refreshToken) throw new Error('No hay refresh token');
    const response = await fetch('https://kronobackend-ctach7cya4cxb9d8.spaincentral-01.azurewebsites.net/api/auth/refresh', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ refreshToken: this._refreshToken }),
    });
    if (!response.ok) throw new Error('Error al refrescar token');
    const data = await response.json();
    this._accessToken = data.accessToken;
    this._refreshToken = data.refreshToken;
    const stored = JSON.parse(sessionStorage.getItem('krono_session') || '{}');
    stored.accessToken = this._accessToken;
    stored.refreshToken = this._refreshToken;
    sessionStorage.setItem('krono_session', JSON.stringify(stored));
  }

  /**
   * Cierra la sesión: borra tokens y datos del usuario.
   */
  logout() {
    this._user = null;
    this._accessToken = null;
    this._refreshToken = null;
    sessionStorage.removeItem('krono_session');
    window.dispatchEvent(new Event('session-changed'));
  }

  /** Guarda la sesión en sessionStorage para persistencia entre recargas */
  _save() {
    sessionStorage.setItem('krono_session', JSON.stringify({
      user: this._user,
      accessToken: this._accessToken,
      refreshToken: this._refreshToken,
    }));
  }

  /** Carga la sesión desde sessionStorage al iniciar la app */
  _load() {
    const raw = sessionStorage.getItem('krono_session');
    if (raw) {
      try {
        const data = JSON.parse(raw);
        this._user = data.user || null;
        this._accessToken = data.accessToken || null;
        this._refreshToken = data.refreshToken || null;
      } catch {
        // Si los datos están corruptos, iniciar sin sesión
        this._user = null;
        this._accessToken = null;
        this._refreshToken = null;
      }
    }
  }
}